feat: add a step template to run custom terraform scripts with packages

alastairtree opened the pull request:
- useful for maintenance tasks like terraform import in runbooks or terraform state upgrades
- follows very similar behaviour as built in terraform apply and terraform plan steps

### Step template checklist

- [x] `Id` should be a **GUID** that is not `00000000-0000-0000-0000-000000000000`
  - **NOTE** If you are modifying an existing step template, please make sure that you **do not** modify the `Id` property *(updating the `Id` will break the Library sync functionality in Octopus)*. 
- [x] `Version` should be incremented, otherwise the integration with Octopus won't update the step template correctly
- [x] Parameter names should not start with `$`
- [x] **Step template parameter names (the ones declared in the JSON, not the script body) should be prefixed with a namespace so that they are less likely to clash with other user-defined variables in Octopus** (see [this issue](https://github.com/OctopusDeploy/Issues/issues/2126)). For example, use an abbreviated name of the step template or the category of the step template).
- [x] `LastModifiedBy` field must be present, and (_optionally_) updated with the correct author
- [x] If a new `Category` has been created:
   - [x] An image with the name `{categoryname}.png` must be present under the `step-templates/logos` folder
   - [x] The `switch` in the `humanize` function in [`gulpfile.babel.js`](https://github.com/OctopusDeploy/Library/blob/master/gulpfile.babel.js#L92) must have a `case` statement corresponding to it

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/OctopusDeploy/Library?pullRequest=1076) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/OctopusDeploy/Library?pullRequest=1076) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OctopusDeploy/Library?pullRequest=1076) it.</sub>

paulegradie commented on step-templates/terraform-custom-script.json:
```diff
@@ -0,0 +1,109 @@
+        "Name": "InfrastructurePackage",
+        "Id": "582b5176-aa5d-4166-a5fa-9b760a3b2d7d",
+        "PackageId": null,
+        "FeedId": "feeds-builtin",
```
You can set the FeedId to null. If its set to a non-variable value, it will cause a validation error when imported and copied into a non-default space. See https://github.com/OctopusDeploy/Issues/issues/6733

paulegradie reviewed (changes requested):
Thanks for this. Just one item on the Package Feed Id. See issue https://github.com/OctopusDeploy/Issues/issues/6733

We are currently performing a similar fix on the other templates. For reference: https://github.com/OctopusDeploy/Library/pull/1077

alastairtree commented:
Thanks. Fixed.

paulegradie commented on step-templates/terraform-custom-script.json:
```diff
@@ -0,0 +1,109 @@
+{
+    "Id": "1d2a62d0-1f12-4417-8566-4ca1b8e6a69c",
+    "Name": "Execute Custom Terraform Script with Package",
+    "Description": "Run a custom terraform script using a package and an Azure Account.\n\nE.g. run \"terraform taint some-resource\" in the context of your terraform package files and authenticated against Azure using credentials in an Octopus Azure Account variable. This step operates very similarly to the built-in Terraform Apply step but allowing you the control to specift custom terraform co ...
```
There is a typo here in 'specift'. 
```suggestion
    "Description": "Run a custom terraform script using a package and an Azure Account.\n\nE.g. run \"terraform taint some-resource\" in the context of your terraform package files and authenticated against Azure using credentials in an Octopus Azure Account variable. This step operates very similarly to the built-in Terraform Apply step but allowing you the control to specify custom terraform commands. Very useful for runbooks and terraform import.",
```

paulegradie reviewed (dismissed):
Thanks for your contribution @alastairtree. Just one last typo to correct.  Thanks also for nulling the feed Id.

@harrisonmeister This looks okay to me.

harrisonmeister commented on step-templates/terraform-custom-script.json:
```diff
@@ -0,0 +1,109 @@
+    "Parameters": [
+      {
+        "Id": "0170e094-b282-4924-b1ed-5103015e366f",
+        "Name": "TerraformPackageId",
```
I can't see this Parameter referenced in the script? Was it by chance replaced with the `InfrastructurePackage` parameter? If so, I think the `TerraformPackageId` parameter can be safely removed.

alastairtree replied:
The package variable is required because the step has a "Referenced Package" so that the project can select a package and then it is referenced in the script as InfrastructurePackage.

![image](https://user-images.githubusercontent.com/6273429/111202302-9d567680-85bb-11eb-8434-c99601983d80.png)

harrisonmeister replied:
Thanks @alastairtree - not sure how I missed it referenced: `"PackageParameterName": "TerraformPackageId"` :smile:

harrisonmeister reviewed (commented):
One query on the use of the `TerraformPackageId`. Apart from that, it LGTM

DevOpsDerek commented:
Hi @alastairtree, are you ok with @harrisonmeister's suggestions?

alastairtree commented:
I have added a comment that should resolve the query by @harrisonmeister. Think he missed the use of a referenced package. Have also updated the PR with the latest changes which just make the init step optional so that you can do init manually if you require. This should be ready to merge I think.

harrisonmeister reviewed (approved):
LGTM

alastairtree commented:
I cannot see this on my Octopus instance - is there any other release process required?

harrisonmeister commented:
@alastairtree 

> I cannot see this on my Octopus instance - is there any other release process required?

Stupid question, but are Community step templates enabled on your Octopus instance and have synced recently?
![image](https://user-images.githubusercontent.com/1418993/112336209-cb6f4100-8cb4-11eb-844d-7257bab8365d.png)

I've checked my local octopus instance and can see the step:

![image](https://user-images.githubusercontent.com/1418993/112336347-ec379680-8cb4-11eb-8553-3bc89711bc2b.png)

alastairtree commented:
Yes they are, but sorry I am an idiot, I have already installed that template so it does not appear when searching for it. Apologies

harrisonmeister commented:
> Yes they are, but sorry I am an idiot, I have already installed that template so it does not appear when searching for it. Apologies

No worries at all!
